pallets/otc-settlements/src/lib.rs

```diff
@@ -0,0 +1,542 @@
+
+		// // Compare OTC and Router price
+		let router_price_after = T::Router::spot_price(&route).unwrap();
+		log::debug!(
+			target: "offchain_worker::settle_otc",
+			"final router price: {:?}   otc_price: {:?} ",
+			router_price_after,
+			otc_price
+		);
+
+		// in the case of fully fillable orders, the resulting price is not important
+		if otc.partially_fillable {
+			let price_diff = {
+				if otc_price > router_price_after {
+					otc_price.saturating_sub(router_price_after)
+				} else {
+					router_price_after.saturating_sub(otc_price)
+				}
+			};
+
+			let price_precision = T::PricePrecision::get()
+				.checked_mul(&otc_price)
+				.ok_or(ArithmeticError::Overflow)?;
+			if price_diff > price_precision {
+				ensure!(router_price_after <= otc_price, Error::<T>::TradeAmountTooHigh);
+				ensure!(router_price_after >= otc_price, Error::<T>::TradeAmountTooLow);
+			}
+		}
+
+		let asset_a_balance_after_router_trade = <T as Config>::Currency::free_balance(asset_a, &pallet_acc);
+
+		let profit = asset_a_balance_after_router_trade
+			.checked_sub(amount)
+			.ok_or(ArithmeticError::Overflow)?;
+
+		Self::ensure_min_profit(asset_a, profit)?;
+
+		<T as Config>::Currency::transfer(asset_a, &pallet_acc, &T::ProfitReceiver::get(), profit)?;
+
+		<T as Config>::Currency::withdraw(asset_a, &pallet_acc, amount)?;
+
+		let asset_a_balance_after = <T as Config>::Currency::free_balance(asset_a, &pallet_acc);
+		let asset_b_balance_after = <T as Config>::Currency::free_balance(asset_b, &pallet_acc);
+
+		ensure!(
+			asset_a_balance_after == asset_a_balance_before,
+			Error::<T>::InvalidConditions
+		);
+		ensure!(
+			asset_b_balance_after == asset_b_balance_before,
+			Error::<T>::InvalidConditions
+		);
+
+		Self::deposit_event(Event::Executed {
+			otc_id,
+			otc_asset_in: asset_a,
+			otc_asset_out: asset_b,
+			otc_amount_in: amount,
+			otc_amount_out,
+			trade_amount_in: otc_amount_out,
+			trade_amount_out: amount + profit,
+		});
+
+		Ok(())
+	}
+
+	/// Store the latest block number in the offchain storage.
+	/// Returns `true` if `block_number` is newer than the block number stored in the storage.
+	fn try_update_last_block_storage(block_number: BlockNumberFor<T>) -> bool {
+		let last_update_storage = StorageValueRef::persistent(OFFCHAIN_WORKER_DATA_LAST_UPDATE);
+		let last_update = last_update_storage
+			.get::<BlockNumberFor<T>>()
+			.unwrap_or_default()
+			.unwrap_or_default();
+
+		if block_number > last_update {
+			last_update_storage.set(&block_number);
+			true
+		} else {
+			false
+		}
+	}
+
+	/// Sort open OTCs orders and save a list in the offchain storage.
+	fn sort_otcs(block_number: BlockNumberFor<T>) {
+		log::debug!(
+			target: "offchain_worker::sort_otcs",
+			"sort_otcs()");
+
+		// acquire offchain worker lock.
+		let lock_expiration = Duration::from_millis(LOCK_DURATION);
+		let mut lock = StorageLock::<'_, Time>::with_deadline(OFFCHAIN_WORKER_LOCK, lock_expiration);
```

Comment:
shouldn't you return the lock after sorting finishes? you want to make sure the array is not sorted again sooner than in 5 sec? if you want to make sure the otcs are sorted only once per block, you should cache it based on block height & hash - same block should yield the same order right?

Reply:
The lock guard should be dropped at the end of `sort_otcs()` and the lock is unlocked. Also the list is updated every time there is a newer block (`try_update_last_block_storage()`).